### PR TITLE
Multi-monitor option

### PR DESCRIPTION
**Pull request by franglais125**
Hi!

I implemented an option to remove the top bar only on the main monitor. I made it optional since a lot of people want *both* behaviors.

This is the reason the pull looks big: I had to add a few files:
- Settings.ui
- prefs.js
- utils.js (I kept it separate from util.js)
- and the schema

Question about the "About" tab I included: I cite the license a being GPL2 (as this is required when uploading extensions to e.g.o), but on your github the license is MIT. We can change this to whatever you like.

I'll add two screenshots soon.

### CODE REVIEW

**Comment by franglais125** [comment]
screenshot from 2017-04-11 13-54-50
screenshot from 2017-04-11 13-55-07

**Comment by deadalnix** [comment]
Is there a reason to introduce settings for this ? I mean we never ever want to have this triggering on another monitor.

**Comment by MannyC** [comment]
If you look at https://github.com/deadalnix/pixel-saver/issues/80#issuecomment-240309667 some people want it to be an option as they're happy with how it works currently.   Adding in the settings page obviously adds a bit of code, but you're going to ruffle some feathers if it isn't configurable.

**Comment by franglais125** [comment]
@deadalnix Exactly, when I was working on this I noticed the same comment that @MannyC referenced, but I also received other comments (@antmak) asking me to make it configurable.

It's ultimately up to you. If you don't want to make it configurable, you can cherry-pick the first commit in this PR!

**Comment by kmhofmann** [comment]
I can confirm that this works like a charm and is highly useful. This makes the extension usable on my 3-monitor setup.
I think this should be enabled by default, and I see the toggle option purely as a bonus for people who want to switch it off (but why would you?).

**Comment by antmak** [comment]
@kmhofmann

>I think this should be enabled by default,

No, wrong way. I use pixel-saver on several 2-monitor configurations under Fedora and CentOS. This option is harmful for me, and I wanna simple sane default: save my pixels. Window titles are evil. Hide this option somewhere deeper.

**Comment by MannyC** [comment]
I think @antmak demonstrates the wish/need to make this optional.  Personally I vote for the enabled by default option as currently pixel-saver does not act sanely for me on multiple monitors.

**Comment by franglais125** [comment]
Disclaimer: I don't have strong feelings, as I don't use this extension. I implemented this as I thought it would be something cool to have, nothing more.

The default for the option is `false`, but it can obviously be changed very easily. Also, the fact that it's an option should make everyone happy (I hope?), as no one is forced to use the extension in a given way.

@deadalnix is there anything else you'd like to see discussed/added?

**Comment by kmhofmann** [comment]
@antmak It is far from obvious for the common user (or first-time Gnome user, as I was) how to un-maximize an already maximized window on a secondary monitor when the top bar isn't there. Heck, I used xkill until I looked up better alternatives. :)

The option should be enabled by default, purely from a "good UX for everyone" perspective.
It's not a problem for anyone who wants to save a few more pixels to switch this off.

**Comment by antmak** [comment]
@kmhofmann pixel-saver is not default plugin in gnome. 

Let's return a window title to the main monitor, dont be shy... Oh, just turn off pixel-saver :)

File: pixel-saver@deadalnix.me/Settings.ui [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+          <property name="expand">False</property>
+          <property name="fill">True</property>
+          <property name="position">5</property>
+        </packing>
+      </child>
+      <child>
+        <object class="GtkLabel" id="GNU_label">
+          <property name="valign">end</property>
+          <property name="label" translatable="yes">&lt;span size="small"&gt;This program comes with ABSOLUTELY NO WARRANTY.
+Sethe &lt;a href="https://www.gnu.org/licenses/old-licenses/gpl-2.0.html"&gt;GNU General Public License, version 2 or later&lt;/a&gt; for details.&lt;/span&gt;</property>
```
Comment (zapashcanon) [inline pixel-saver@deadalnix.me/Settings.ui]: @franglais125 I think something's wrong here: `Sethe`. Spotted looking at your screen-shot.

**Comment by deadalnix** [comment]
I put some thought into this. It's clear to me that the default behavior should be to disable pixel saver on the secondary monitor. The expected behavior is that the activity bar merged with the title bar when a window is maximized and there is no activity bar on the secondary monitor.

I'm ok with having an option to keep it enabled anyway as it seems there is a fair amount of demand for it, but that shouldn't be the default behavior.

Is it possible to get a PR with exclusively the disabling of pixel saver on the second monitor ? This would make thing clearer if separate things are separated.

**Comment by franglais125** [comment]
I just rebased against the current master branch.

I set the default to show the titlebars on secondary monitors, and I fixed the typo @zapashcanon found (thanks!).

> Is it possible to get a PR with exclusively the disabling of pixel saver on the second monitor ? This would make thing clearer if separate things are separated.

The first commit does exactly this, if I understand correctly your question.
- [988c5f0cdc02596e9dbda463bb45e5064bb442e3] show titlebars on secondary monitos
- [98bd88be1842c08387a8bad74637887e280c5157] add a settings framework
- [37c01cb83ebf37752fd43d2d12dc17db77f5ead0] make use of the setting

Unless I misunderstood the question.

**Comment by deadalnix** [comment]
Ok I'm goin to cherry ick the first commit and then review more thoroughly the rest. The note will need to be updated to reflect that there is now configuration in pixel saver. I wish to provide out of the box what a regular user will expect, but it seems clear that there is demand from power users (who I assume know the keyboard shortcuts to maximize/unmaximize) to save more pixel, so let's add a config.

**Comment by franglais125** [comment]
That sounds fine to me!

I do understand that's quite a bit of code that I'm proposing. The whole settings thing requires the xml code, the schemas and a handler. What I want to mention about these things is that I didn't just "come up" with all the code. I based this on the code used by dash-to-dock (that I also then used on other extensions I maintain).

Let me know if you have more questions about it. Here is a quick summary of the new files:

- Settings.ui: Contains the structure of the settings window
- prefs.js: Builds the settings window (allowing translations if we add locales) and connects the `true` or `false` status of the option key
- schema: presents the options (booleans, strings, whatever is necessary) to the extension system
- utils.js: initializes a settings instance, so that they are available to the extension code

**Comment by jhasse** [comment]
I've just tested this and while the titlebar won't get removed on the secondary monitor, the close buttons are still shown in the top panel even if there's only a maximized window on the secondary monitor.

**Comment by franglais125** [comment]
@jhasse Thanks for spotting that. I included a fix in the latest commit. [562108603d0f4486dc041e921995a6d7c587655b].

@deadalnix since now we have a settings framework, we can easily make the buttons optional and fix #75.

File: pixel-saver@deadalnix.me/buttons.js [hunk 2]
```diff
@@ -209,6 +209,9 @@ function updateVisibility() {
 		let win = Util.getWindow();
 		if (win) {
 			visible = win.decorated;
+                        // If still visible, check if on primary monitor
+		        if (visible && settings.get_boolean('only-main-monitor'))
+			        visible = win.is_on_primary_monitor();
```
Comment (jhasse) [inline pixel-saver@deadalnix.me/buttons.js]: You should tabs instead of spaces here, like in the rest of the file.

**Comment by jhasse** [comment]
> @jhasse Thanks for spotting that. I included a fix in the latest commit. [5621086].

Thanks :) One minor thing left: The window title is still shown on the top bar if a window is maximized on a secondary monitor.

File: pixel-saver@deadalnix.me/buttons.js (hunk 2 above)
Comment (franglais125) [inline pixel-saver@deadalnix.me/buttons.js]: Fixed in [9e4987aa2a7051437efca7945367e4627ab04e15].

**Comment by franglais125** [comment]
> Thanks :) One minor thing left: The window title is still shown on the top bar if a window is maximized on a secondary monitor.

For some reason I missed this comment... I just saw it!

I noticed this, but I left it in as I thought it wouldn't bother. In what case is it bad? If an app is on the secondary monitor (without having pixel saver), the app menu shows up anyway.

Is it in case we have two windows of the same app each on a different monitor?

**Comment by deadalnix** [comment]
Ok can someone with multiple monitor can attest what's the state of this PR ? It's a bit difficult on my side for now because I do not have a multi monitor set up to test this.

**Comment by jhasse** [comment]
> I noticed this, but I left it in as I thought it wouldn't bother. In what case is it bad? If an app is on the secondary monitor (without having pixel saver), the app menu shows up anyway.

But if you have two windows of the same app maximized on both monitors (on the primary monitor first), the one on the secondary is displayed two times (in Gtk's top panel and in the top bar) and the title of the window on the primary monitor is not displayed.

Hope you get what I mean :D Maybe it isn't worth fixing.

> Ok can someone with multiple monitor can attest what's the state of this PR ? It's a bit difficult on my side for now because I do not have a multi monitor set up to test this.

I'm using it and it works fine :) Also there's the option to turn it off.

**Comment by franglais125** [comment]
@jhasse I included a fix for the title issue in [45d878fc8b3b0c11c91e554824a0986732dcaac6].

I also added [072d09c8172cee769dc356aa7c70742ae033e85b] to better refresh the appMenu and the buttons! Let me know how it goes...

**Comment by gazhay** [comment]
I use this to fullscreen kodi on a 3rd monitor - if you use fullscreen in kodi it always picks the primary monitor. If you maximise it has a titlebar. This extension helps. 
I would like to see settings introduced as a per application basis.

**Comment by franglais125** [comment]
@gazhay  I'd love to add more settings, but first we need to introduce a settings framework, which this PR does.

**Comment by manuee** [comment]
Love to see action on this - If I may make a suggestion, I believe if we forgot about primary monitor, and instead make this option be for enabling _only whether the monitor has a panel or not_ it would make this more versatile. 
Think of using this extension along with the [Multiple Monitor Panels](https://extensions.gnome.org/extension/323/multiple-monitor-panels/) extension, both together would work lovely if we focused on the existence of a panel on the current monitor =)

**Comment by Beanow** [comment]
Exactly what I needed.
- :+1: as default behavior.
- Integrates well with for example "Dash to panel" as per screenshot.

Since removing the extension does not restore default behavior, logging out is required to test, after that it worked as expected for me.

screenshot from 2017-07-10 17-57-52

**Comment by franglais125** [comment]
It'd be great to see some progress in some of the PRs, not just this one.

At this point I'm considering uploading my own fork to e.g.o, which is only a shame, as I certainly don't want to maintain another extension, nor duplicate efforts.

Unfortunately, given the current status quo, many features are being held back.

**Comment by jhasse** [comment]
@franglais125 I've already created a fork here: https://github.com/jhasse/gnome-shell-extension-pixel-saver-ng
This is because of a bug in Mutter which was fixed in GNOME 3.24, so my fork only works with 3.24 or newer. I could merge your changes there and then upload it to e.g.o if you want.

**Comment by franglais125** [comment]
Thanks for the offer! I'll take look at your changes tomorrow. I just happened to upload to e.g.o! Since I had to push the changes to master on my github, I closed this PR. If not I'd be spamming everyone with every single commit I push.

It's still in review obviously. https://extensions.gnome.org/review/7144

**Pull request closed without merging**